torch/csrc/autograd/record_function.cpp

```diff
@@ -4,20 +4,23 @@
 namespace torch { namespace autograd { namespace profiler {
 
 namespace {
-bool has_callbacks = false;
+RecordFunctionState record_function_state = RecordFunctionState::NO_CALLBACKS;
 std::vector<RecordFunctionCallback> start_callbacks;
 std::vector<RecordFunctionCallback> end_callbacks;
+std::vector<bool> needs_inputs_flags({false});
 thread_local RecordFunction* thread_local_func_ = nullptr;
 }
 
-void pushCallback(RecordFunctionCallback start, RecordFunctionCallback end) {
+void pushCallback(
+    RecordFunctionCallback start, RecordFunctionCallback end, bool needs_inputs) {
   start_callbacks.push_back(start);
   end_callbacks.push_back(end);
-  has_callbacks = true;
-}
-
-void pushCallback(RecordFunctionCallback start) {
-  pushCallback(start, [](const RecordFunction&){});
+  needs_inputs_flags.push_back(needs_inputs_flags.back() || needs_inputs);
+  if (needs_inputs_flags.back()) {
```

Comment:
I am not a fan of this de-normalized logic. There are is enough information in the data structures to always figure out the record_function_state, but pushCallback and pop are trying to maintain it. It would be easier as just:
```
bool  has_callbacks = ...;
size_t call_backs_need_inputs; // ++ on need, -- on popping something that doesn't have need.
```

Then the guard and the API can be more self-documenting;

```
#define RECORD_FUNCTION(fn, inputs, ...)
RecordFunction guard; // does nothing
if (has_callbacks) {
  guard.setName(<name_information>, __VA_ARGS__); // sets initialized_;
                                                                                            // VA_ARGS holds seq_nr if present
  if(call_backs_need_inputs) {
    guard.setInputs(inputs)  
  }
}
```

Reply:
I also feel like we could make `RecordFunction` hold a `optional<CallbackGuard>` field, and then we wouldn't have to worry too much about maintenance of `initialized_` -- you either have the struct which *always* calls before/after callbacks, or you don't.